README.md
#Introduction
Particle swarm optimization (PSO) is a derivative-free global optimum solver. It is inspired by the surprisingly organized behaviour of large groups of simple animals, such as flocks of birds, schools of fish, or swarms of locusts. The individual creatures, or "particles", in this algorithm are primitive, knowing only four simple things: 1 & 2) their own current location in the search space and fitness value, 3) their previous personal best location, and  4) the overall best location found by all the particles in the "swarm". There are no gradients or Hessians to calculate. Each particle continually adjusts its speed and trajectory in the search space based on this information, moving closer towards the global optimum with each iteration. As seen in nature, this computational swarm displays a remarkable level of coherence and coordination despite the simplicity of its individual particles.

Ease of Use
If you are already using the Genetic Algorithm (GA) included with MATLAB's Global Optimization Toolbox, then this PSO toolbox will save you a great deal of time. It can be called from the MATLAB command line using the same syntax as the GA, with some additional options specific to PSO. This will allow a high degree of code re-usability between the PSO toolbox and the GA toolbox. Certain GA-specific parameters such as cross-over and mutation functions will obviously not be applicable to the PSO algorithm. However, many of the commonly used options for the Genetic Algorithm Toolbox may be used interchangeably with PSO since they are both iterative population-based solvers. See >> help pso (from the ./psopt directory) for more details.

#Features
  * NEW: support for distributed computing using MATLAB's parallel computing toolbox.
  * Full support for bounded, linear, and nonlinear constraints.
  * Modular and customizable.
  * Binary optimization. See PSOBINARY function for details.
  * Vectorized fitness functions.
  * Solver parameters controlled using 'options' structure similar to existing MATLAB optimization solvers.
  * User-defined custom plots may be written using same template as GA plotting functions.
  * Another optimization solver may be called as a "hybrid function" to refine PSO results.

A demo function is included, with a small library of test functions. To run the demo, from the psopt directory, call >> psodemo with no inputs or outputs.

New features and bug fixes will continue to be released until this is made redundant by the release of an official MATLAB PSO toolbox. Bug reports and feature requests are welcome.

Special thanks to the following people for contributing code and bug fixes:
  * Ben Xin Kang of the University of Hong Kong
  * Christian Hansen of the University of Hannover
  * Erik Schreurs from the MATLAB Central community
  * J. Oliver of Brigham Young University
  * Michael Johnston of the IRIS toolbox
  * Ziqiang (Kevin) Chen

#Bibliography
  * J Kennedy, RC Eberhart, YH Shi. Swarm Intelligence. Academic Press, 2001.
  * Particle Swarm Optimization. http://en.wikipedia.org/wiki/Particle_swarm_optimization
  * RE Perez, K Behdinan. Particle swarm approach for structural design optimization. Computers and Structures 85 (2007) 1579–1588.
  * SM Mikki, AA Kishk. Particle Swarm Optimization: A Physics-Based Approach. Morgan & Claypool, 2008.

#Addendum A
Nonlinear inequality constraints in the form c(x) ≤ 0 and nonlinear equality constraints of the form ceq(x) = 0 have now been fully implemented. The 'penalize' constraint boundary enforcement method is now default. It has been redesigned and tested extensively, and should work with all types of constraints.

See the following document for the proper syntax for defining nonlinear constraint functions: http://www.mathworks.com/help/optim/ug/writing-constraints.html#brhkghv-16.
To see a demonstration of nonlinear inequality constraints using a quadrifolium overlaid on Rosenbrock's function, run PSODEMO and choose 'nonlinearconstrdemo' as the test function.

#Addendum B
See the following guide in the GA toolbox documentation to get started on using the parallel computing toolbox.
http://www.mathworks.com/help/gads/genetic-algorithm-options.html#f17234

#Addendum C
If you are a beginner hoping to learn to use this toolbox for work or school, here are some essential readings:
  * MATLAB's Optimization Toolbox: http://www.mathworks.com/help/optim/index.html
  * MATLAB's Global Optimization Toolbox: http://www.mathworks.com/help/gads/index.html
  * MATLAB's Genetic Algorithm: http://www.mathworks.com/help/gads/genetic-algorithm.html

#Addendum D
There is now a particle swarm optimizer included with the Global Optimization Toolbox. It does not seem to handle constraints at this time. If you have a recent version of the Global Optimization Toolbox installed, you will need to set the path appropriately in your code to use this toolbox.

releasenotes.txt
Development Notes for psopt toolbox

Future plans (in no particular order):
* Performance improvement: Automatically check for existence of constraints, skip boundary-checking if unconstrained.
* Performance improvement: Automatically vectorize fitness functions supplied by user, if not already vectorized.
* Performance improvement: Detect and eliminate dependent (i.e. redundant) linear constraints.
* Performance improvement: Automatically choose "penalize" versus "soft" constraint handling method based on type of constraints in problem. If any equality constraints exist, using "penalize", otherwise default to "soft". If "soft", automatically switch to "penalize" if more than a given percentage of particles (initial or interim) are infeasible.
* Feature: An output function defined by options.OutputFcns that can save current swarm state in a file which can be reloaded later as a problem structure, in case of a system crash.
* Feature: Implement multiobjective optimization option; find Pareto front.
* Feature: A better-looking psodemo function.
* Feature: Ability to define neighborhoods of particles to determine non-local best positions (social parameter).
  - Social stereotyping using clustering algorithm (Kennedy 2000)
  - Von Neumann topology (Kennedy 2002)
* Feature: Ability to solve problems with discrete parameters
* Feature: New convergence criterion: isothermal equilibrium (Mikki 2008, Chapter 3).
* Feature: Clerc's method: constriction factor chi to prevent explosive divergence of swarm.
* Feature: Swarm coherence and collision-avoidance parameters?
* Bug fix: "Red Flag" patch object not showing up in PSODEMO when run with surface plot demonstration.

ver 1.31.2 (2016-02-25)
* Bug fix: added appropriate "end" lines to nested functions in PSOITERATE (contributed by unknown user via Google Code)
* Update: MATLABPOOL has been deprecated in recent versions of MATLAB. Updated distributed computing code as appropriate.
* New version numbering scheme.

ver 20130430
* Bug fix: corrected typo in PSOCHECKPOPULATIONINITRANGE which was causing incorrect handling of initial particle distribution boundaries. Thanks to Erik from the MATLAB Central community.

ver 20131220
* Performance improvement: Added parallel computing capability to constraint violation calculations for "penalize" method.
* Performance improvement: Optimized and refactored code for calculating constraint violations for "penalize" method.
* Performance improvement: Added parallel computing capability for determining constraint violations for "soft" method.
* Minor bugfix: in psoboundspenalize, code for applying tolerances to equality constraints will now properly execute after checking for existence of equality constraints.

ver 20130702:
* Bug fix: multiple errors in variable name when checking for verbosity before warnings were thrown -- did not properly specify the options structure array. Thanks to Christian Hansen of the University of Hannover for pointing this out.
* Documentation updated for PSOOPTIMSET function.

ver 20130615:
* Added parallel computing capability. Once the distributed computing worker pool has been appropriately set up using MATLABPOOL, run PSO with options.UseParallel set to 'always'. If both Vectorized and UseParallel options are set to 'on' and 'always', respectively, then the vectorized option will override the parallel option. If PSO is unable to find an installed version of the Parallel Computing Toolbox, then the problem will quietly switch to serial computation. Thanks to Michael Johnston and J. Oliver for their suggestions and code contributions.
* Most warning messages will now be suppressed unless options.Display is set 'diagnose'.
* Changed exit flag for reaching total time limit from 5 to -5, to match exit flags for GA.
* Implemented stall time limit option, with the algorithm terminating with exit flag -4 if there has been no improvement in global best fitness after a certain number of seconds. Default value is Inf.

ver 20130515b:
* Fixed namespace issues with duplicate copies of the psoiterate.m file, one of which was outdated.

ver 20130515:
* Merged back changes from version 20100818 which did not make it into ver 20130502
* Fixed error with constraint enforcement method set to 'penalize' with an unconstrained problem.
* Slight modification to penalty calculation method to prevent infeasible points from being selected over feasible points when selecting the global best.
* Default constraint enforcement method is now 'penalize'. Since this is a relatively new feature, please double-check the feasibility of solutions to your constrained problems before accepting results for mission-critical applications.
* Documentation updated in PSOOPTIMSET help file.
* Plotting axes should no longer cause errors if options.PlotInterval > 1 (suggestion made by Ziqiang (Kevin) Chen, 13 Nov 2012)

ver 20130502:
* Added a new constraint handling method: 'penalize', described in Perez 2007. Both nonlinear inequality and equality constraints can now be handled gracefully by this toolbox. It also works well for handling linear and bounded constraints. If there are no problems reported after a certain amount of time "in the wild", this will become the default constraint handling method.

ver 20100818:
* Added a note in the documentation suggesting that social and cognitive attraction parameters should have a sum of less than 4, as demonstrated by Perez 2007. A warning message appears (but the optimization continues) if the user defines an unstable swarm.
* A time limit (in seconds) can now be set using the 'TimeLimit' option in PSOOPTIMSET. The algorithm will attempt to estimate the average amount of time per iteration, and end the optimization, returning its best results so far, when the next iteration is predicted to exceed the time limit.

psooptimset.m
function options = psooptimset(varargin)
% Creates an options structure for pso.
%
% Syntax:
% psooptimset
% options = psooptimset
% options = psooptimset(@pso)
% options = psooptimset(@psomultiobj)
% options = psooptimset('param1',value1,'param2',value2,...)
% options = psooptimset(oldopts,'param1',value1,...)
% options = psooptimset(oldopts,newopts)
%
% Description:
% psooptimset with no input or output arguments displays a complete list of
% parameters with their valid values.
% 
% options = psooptimset (with no input arguments) creates a structure
% called options that contains the options, or parameters, for the pso
% algorithm and sets parameters to [], indicating default values will be
% used.
% 
% options = psooptimset(@pso) creates a structure called options that
% contains the default options for the genetic algorithm.
% 
% options = psooptimset(@psomultiobj) creates a structure called options
% that contains the default options for psomultiobj. Not yet implemented
% 
% options = psooptimset('param1',value1,'param2',value2,...) creates a
% structure options and sets the value of 'param1' to value1, 'param2' to
% value2, and so on. Any unspecified parameters are set to their default
% values. Case is ignored for parameter names.
% 
% options = psooptimset(oldopts,'param1',value1,...) creates a copy of
% oldopts, modifying the specified parameters with the specified values.
% 
% options = psooptimset(oldopts,newopts) combines an existing options
% structure, oldopts, with a new options structure, newopts. Any parameters
% in newopts with nonempty values overwrite the corresponding old
% parameters in oldopts.
%
% Again, type psooptimset with no input arguments to display a list of
% options which may be set.
%
% NOTE regarding the ConstrBoundary option:
% The 'penalize' constraint management option as described in Perez 2007 is
% the default since it seems to provide the best combination of performance
% and versatility. It combines the best of the 'soft' and 'absorb' methods
% and adds a penalty calculated for every infeasible point in the search
% space.
%
% Although the 'penalize' method is preferred, the 'soft' or 'absorb' style
% boundaries may still be used. They are described below. They are not very
% effective if your problem contains any nonlinear equality constraints of
% the form ceq(x) = 0 or linear equality constraints of the form Aeq*x =
% beq. If your problem does not require any equality constraints, then
% calculations may be faster with 'soft' boundaries since infeasible
% solutions are not evaluated at all and simply assigned a fitness score of
% infinity.
%
% A 'soft' boundary allows particles to leave problem bounds, but sets
% their fitness scores to Inf if they do. This can save time, since
% infeasible points are not evaluated. This will also save you some trouble
% if your fitness function has the possibility of throwing errors if it
% tries to evaluate an infeasible point in the search space.
% 
% The 'reflect' and 'absorb' options prevent the particles from travelling
% outside the problem bounds at all. Note that 'reflect' has only been
% implemented for bounded constraints, and is not really suitable for any
% other problem type at this time. The 'absorb' method may suffer from poor
% performance if linear or nonlinear equality constraints are used. Both
% are likely to be deprecated in the near future.
%
% NOTE regarding cognitive and social attraction parameters:
% Perez and Behdinan (2007) demonstrated that the particle swarm is only
% stable if the following conditions are satisfied:
% Given that C0 = particle inertia
%            C1 = options.SocialAttraction
%            C2 = options.CognitiveAttraction
%    1) 0 < (C1 + C2) < 4
%    2) (C1 + C2)/2 - 1 < C0 < 1
% If conditions 1 and 2 are satisfied, the system will be guaranteed to
% converge to a stable equilibrium point. However, whether or not this
% point is actually the global minimum cannot be guaranteed, and its
% acceptability as a solution should be verified by the user.
%
% Bibliography
% RE Perez and K Behdinan. "Particle swarm approach for structural
% design optimization." Computers and Structures, Vol. 85:1579-88, 2007.
%
% See also:
% pso, psodemo

% Default options
options.CognitiveAttraction = 0.5 ;
options.ConstrBoundary = 'penalize' ; 
options.AccelerationFcn = @psoiterate ;
options.DemoMode = 'off' ;
options.Display = 'final' ;
options.FitnessLimit = -inf ;
options.Generations = 200 ;
options.HybridFcn = [] ;
options.InitialPopulation = [] ;
options.InitialVelocities = [] ;
options.KnownMin = [] ;
options.OutputFcns = {} ;
options.PlotFcns = {} ;
options.PlotInterval = 1 ;
options.PopInitRange = [0;1] ;
options.PopulationSize = 40 ;
options.PopulationType = 'doubleVector' ;
options.SocialAttraction = 1.25 ;
options.StallGenLimit = 50 ;
options.StallTimeLimit = Inf ;
options.TimeLimit = Inf ;
options.TolCon = 1e-6 ;
options.TolFun = 1e-6 ;
options.UseParallel = 'never' ;
options.Vectorized = 'off' ;
options.VelocityLimit = [] ;

if ~nargin && ~nargout
    fprintf('\n')
    fprintf('Available options for PSOOPTIMSET {defaults in braces}:\n\n')
    fprintf('    AccelerationFcn: [Function handle | {@psoiterate}]\n') ;
    fprintf('CognitiveAttraction: [Positive scalar | {%g}]\n',...
        options.CognitiveAttraction) ;
    fprintf('     ConstrBoundary: [''soft'' | ''penalize'' | ''reflect'' | ''absorb'' | {''%s''}]\n',...
        options.ConstrBoundary) ;
    fprintf('            Display: [''off'' | ''final'' | ''diagnose'' | {''%s''}]\n',...
        options.Display) ;
    fprintf('           DemoMode: [''fast'' | ''pretty'' | ''on'' | ''off'' | {''%s''}]\n',...
        options.DemoMode) ;
    fprintf('       FitnessLimit: [Scalar | {%g}]\n',...
        options.FitnessLimit) ;
    fprintf('        Generations: [Positive integer | {%g}]\n',...
        options.Generations) ;
    msg = sprintf('          HybridFcn: [@fminsearch | @patternsearch |');
    fprintf('%s @fminunc | @fmincon | {[]}]\n',msg)
    fprintf('  InitialPopulation: [empty matrix | nxnvars matrix | {[]}]\n')
    fprintf('  InitialVelocities: [empty matrix | nxnvars matrix | {[]}]\n')
    % PlotFcns, a bit tricky to turn into a string:
    % ---------------------------------------------------------------------
    if ~isempty(options.PlotFcns)
        msg = '{' ;
        for i = 1:length(options.PlotFcns)
            msg = sprintf('%s@%s, ',msg,func2str(options.PlotFcns{i})) ;
        end % for i
        msg = sprintf('%s\b\b}',msg) ;
    else
        msg = '{}' ;
    end
    fprintf('           PlotFcns: [Cell array of fcn handles | {%s}]\n',...
        msg) ;
    % ---------------------------------------------------------------------
    fprintf('       PlotInterval: [Positive integer | {%g}]\n',...
        options.PlotInterval) ;
    fprintf('       PopInitRange: [2x1 vector | 2xnvars matrix | {%s}]\n',...
        mat2str(options.PopInitRange)) ;
    fprintf('     PopulationSize: [Positive integer | {%g}]\n',...
        options.PopulationSize) ;
    fprintf('     PopulationType: [''bitstring'' | ''doubleVector'' | {''%s''}]\n',...
        options.PopulationType) ;
    fprintf('   SocialAttraction: [Positive scalar | {%g}]\n',...
        options.SocialAttraction) ;
    fprintf('      StallGenLimit: [Positive integer | {%g} ]\n',...
        options.StallGenLimit) ;
    fprintf('     StallTimeLimit: [Positive scalar (seconds) | {%g} ]\n',...
        options.StallTimeLimit) ;
    fprintf('          TimeLimit: [Positive scalar (seconds) | {%g} ]\n',...
        options.StallGenLimit) ;
    fprintf('             TolFun: [Positive scalar | {%g}]\n',...
        options.TolFun) ;
    fprintf('             TolCon: [Positive scalar | {%g}]\n',...
        options.TolCon) ;
    fprintf('        UseParallel: [''always'' | ''never'' | {''%s''}]\n',...
        options.UseParallel) ;
    fprintf('         Vectorized: [''on'' | ''off'' | {''%s''}]\n',...
        options.Vectorized) ;
    fprintf('      VelocityLimit: [Positive scalar | {[]}]\n');
    fprintf('\n')
    clear options
    return
end

if ~nargin || isequal(varargin{1},@pso)
    return % Return default values
elseif isstruct(varargin{1})
    oldoptions = varargin{1} ;
    fieldsprovided = fieldnames(oldoptions) ;
    if nargin == 2 && isstruct(varargin{2})
        newoptions = varargin{2} ;
        newfields = fieldnames(newoptions) ;
    end
end

requiredfields = fieldnames(options) ;

% Find any input arguments that match valid field names. If they exist,
% replace the default values with them.
for i = 1:size(requiredfields,1)
    idx = find(cellfun(@(varargin)strcmpi(varargin,requiredfields{i,1}),...
        varargin)) ;
    if ~isempty(idx)
        options.(requiredfields{i,1}) = varargin(idx(end) + 1) ;
        options.(requiredfields{i,1}) = options.(requiredfields{i,1}){:} ;
    elseif exist('fieldsprovided','var')
        fieldidx = find(cellfun(@(fieldsprovided)strcmp(fieldsprovided,...
            requiredfields{i,1}),...
            fieldsprovided)) ;
        if ~isempty(fieldidx)
            options.(requiredfields{i,1}) = ...
                oldoptions.(fieldsprovided{fieldidx}) ;
        end
        if exist('newfields','var')
            newfieldidx = find(cellfun(@(newfields)strcmp(newfields,...
                requiredfields{i,1}),...
                newfields)) ;
            if ~isempty(newfieldidx)
                options.(requiredfields{i,1}) = ...
                    newoptions.(newfields{newfieldidx}) ;
            end
        end
    end % if ~isempty
end % for i

% Some robustness
if isequal(size(options.PopInitRange),[1 2])
    options.PopInitRange = options.PopInitRange' ;
end